File: docs/validation/cpp/app/src/scenarios/s09_type_exports.cpp
// S-09: Public type export surface
//
// Every public type must resolve from the single umbrella header
// (no internal include paths) and be usable, not just nameable.
// See VALIDATION_PLAN.md §8 S-09 and §9.5.
#include "s09_type_exports.hpp"
#include "machine_config/machine_config.hpp"

namespace s09 {

using namespace machine_config;

// OpcuaConfig is never populated by MockConfigBuilder's default build (no
// OPCUA fixture involved here) — type-annotating a never-called parameter is
// exactly what VALIDATION_PLAN.md §8 S-09 allows: "Construct OR
// type-annotate a variable with each type."
static void typeCheckOpcua(const OpcuaConfig&) {}

scenarios::Result run(const std::filesystem::path&, const std::filesystem::path&) {
    (void)&typeCheckOpcua;

    MockConfigBuilder builder;
    builder.laser_count = 2;
    MachineConfig config = builder.build();

    const MachineConfigMeta& meta = config.meta;
    const Machine& machine = config.machine;

    if (config.optical_trains.empty()) {
        return {false, "builder produced zero optical trains"};
    }
    const OpticalTrain& train = config.optical_trains[0];
    const Scanner& scanner = train.scanner;
    const LightSource& lightSource = train.light_source;
    const Collimator& collimator = train.collimator;
    const ScannerCard& scannerCard = train.scanner_card;
    const OptionalComponents& optionalComponents = train.optional_components;

    if (!optionalComponents.clearbox.has_value()) {
        return {false, "expected builder to include a clearbox by default"};
    }
    const ClearBox& clearbox = *optionalComponents.clearbox;

    if (!train.scan_field_correction_file.has_value()) {
        return {false, "expected builder to include an SFCF by default"};
    }
    const ScanFieldCorrectionFile& sfcf = *train.scan_field_correction_file;

    MachineConfigReader reader{std::filesystem::path{"unused"}};
    MachineConfigWriter writer{config};
    (void)reader;
    (void)writer;

    if (meta.machine_name.empty() || machine.manufacturer.empty() ||
        scanner.manufacturer.empty() || lightSource.manufacturer.empty() ||
        collimator.manufacturer.empty() || scannerCard.manufacturer.empty() ||
        clearbox.ip_address.empty() || sfcf.document_name.empty()) {
        return {false, "one or more constructed fields were unexpectedly empty"};
    }

    return {true, "all public types resolve and are usable from the umbrella header"};
}

} // namespace s09

File: docs/validation/rust/app/src/scenarios/s09_type_exports.rs
// S-09: Public type export surface
//
// ID:          S-09
// Title:       Verify all public model types are importable from the library surface
// Category:    happy-path
// Layer:       public API / packaging
// Action:      Import every consumer-facing type directly from the crate root
//              (no `machine_config::models::...` or other sub-module path) and
//              prove each is usable, not just nameable.
// Rationale:   If a consumer must import from an internal path, the library's
//              public API surface is incomplete. See VALIDATION_PLAN.md §8.

use machine_config::{
    Machine, MachineConfig, MachineConfigMeta, MachineConfigReader, MachineConfigWriter,
    MockConfigBuilder, OpticalTrain, Scanner, LightSource, Collimator, ScannerCard,
    ClearBox, ScanFieldCorrectionFile, OpcuaConfig, OptionalComponents,
};
use std::path::Path;

// `MockConfigBuilder::new(n).build()` never sets OPCUA (that only comes from
// the OPCUA fixture used in S-07) — there's no real value to hold here, so
// type-annotating a never-called parameter is what VALIDATION_PLAN.md §8 S-09
// explicitly allows: "Construct OR type-annotate a variable with each type."
#[allow(dead_code)]
fn type_check_opcua(_opcua: &OpcuaConfig) {}

pub fn run(_fixtures_dir: &Path, _real_dir: &Path) -> (bool, String) {
    let builder: MockConfigBuilder = MockConfigBuilder::new(2);
    let config: MachineConfig = builder.build();

    let meta: &MachineConfigMeta = &config.meta;
    let machine: &Machine = &config.machine;

    if config.optical_trains.is_empty() {
        return (false, "builder produced zero optical trains".to_string());
    }
    let train: &OpticalTrain = &config.optical_trains[0];
    let scanner: &Scanner = &train.scanner;
    let light_source: &LightSource = &train.light_source;
    let collimator: &Collimator = &train.collimator;
    let scanner_card: &ScannerCard = &train.scanner_card;
    let optional_components: &OptionalComponents = &train.optional_components;

    let clearbox: &ClearBox = match &optional_components.clearbox {
        Some(cb) => cb,
        None => return (false, "expected builder to include a clearbox by default".to_string()),
    };
    let sfcf: &ScanFieldCorrectionFile = match &train.scan_field_correction_file {
        Some(s) => s,
        None => return (false, "expected builder to include an SFCF by default".to_string()),
    };

    let writer: MachineConfigWriter = MachineConfigWriter::new(&config);
    let _reader_type_check: fn(&str) -> machine_config::error::Result<MachineConfigReader> =
        |p| MachineConfigReader::open(p);

    if meta.machine_name.is_empty()
        || machine.manufacturer.is_empty()
        || scanner.manufacturer.is_empty()
        || light_source.manufacturer.is_empty()
        || collimator.manufacturer.is_empty()
        || scanner_card.manufacturer.is_empty()
        || clearbox.ip_address.is_empty()
        || sfcf.document_name.is_empty()
    {
        return (false, "one or more constructed fields were unexpectedly empty".to_string());
    }
    let _ = writer;

    (
        true,
        "all public types resolve and are usable from the crate root".to_string(),
    )
}
